[REF] v16 revamp - navbar - sri

stefanorigano opened the pull request:
Move the entire navbar-mobile logic from enterprise to community.
Adapt the navbar design to v16 mockup

Requires:
- ~https://github.com/odoo/odoo/pull/83501~. -> merged

See also:
- enterprise: https://github.com/odoo/enterprise/pull/23337


--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

robodoo commented:
[Pull request status dashboard](https://mergebot.odoo.com/odoo/odoo/pull/82434).

stefanorigano commented:
Hello @adr-odoo , here's the PR :)
entrerprise one -> https://github.com/odoo/enterprise/pull/23337

@qsm-odoo , since we moved all `navbar` responsiveness logic from enterprise to community (community uses the `hamburger_menu` now), the navbar's scss that was initially shared between backend/frontend  is now used by website only; I moved the code into the `website` module and applied the necessary adaptations.

@seb-odoo , there are minor changes in `mail` too; the systray badges style is now defined by the navbar, there is no more need to re-define it for the `o_MessagingMenu_counter` component. 

Thanks!

stefanorigano commented:
Thanks for your detailed feedback @seb-odoo ,
 you hilighted the most critic points of this PR indeed...

> However, I find it unfortunate that we are still relying on CSS and font hack in 2022, instead of
just using the SVG directly where we need it, 

I totally agree. Using components would be totally feasible and much easier to maintain compared to a webfont.
Unfortunately our vision is not widely shared, a lot of people find convenient "apply classes" instead.
On the bright side, though, we finally crossed the line of creating and organising our own icons...  of course we are keeping the SVG sources in a separate repo, maybe one day we'll create components for them, who knows ;)

> systray [...] It is currently (and already before your PR) complex to understand what is going on with the style [...] My suggestion for this point would be to give us a clear class to use "systray-icon"

Good point, it was one of my initial tests indeed.  I concluded that it was a bit overkill to define a new class for just an icon that, sometimes, may contain a badge. Looking a the the current selector, thought, I do agree with you... thanks!

----
//  Design glitches:

> the "counter" badge does not have the same height/font/or alignment between discuss menu and activity menu.

I can't replicate, neither in chrome, firefox or safari :/
 
 
 > this counter badge background also changes color,
 
 Yes, it's intentional.
 The aim is to provide better contrast when the navbar is visible and "pave the road" for dark mode.. but that specific color may change (*)
 
 
 > "highlighted" effect etc...
 
 I didn't change this behaviour. Is probably more visible because of the new icons ( that may change too *).
 I do agree btw, either we'll review the behaviour or we adapt the icons.
 
 
 > If we decide that to be the new "style" for activities, maybe we should also change in activity icon in other places, such as chatter
 
 Confirmed, it will be the new pictogram for activities, I'll show you the mockup
 
 
 > the RTC call indicator seems out of plaace
 
 Thanks, I forgot about it
 
 
 (*) Several design aspects/decorations are still temporary

sdegueldre commented:
@stefanorigano unrelated to this PR but you might notice that the badges with notification numbers aren't vertically centered in community only, this seems to come from the use of the "Odoo Unicode Support Noto", I wonder if we could start using Roboto for community as well or simply fix the odoo font?

aab-odoo commented:
Hello @stefanorigano 

I didn't read the diff, but I logged in a runbot to see the result. It looks nice :-)

I found a small glitch as well: in studio, go to the home menu, the 'Notes' and 'Close' buttons aren't in the top right corner as they should (I guess).

stefanorigano commented:
Hello again, thanks for your feedback guys :)

@seb-odoo , you may have a look to 6241e4d59daee80ecb7c4c4544f268a2172d6164, it should solve the opacity misleading behaviour for systray icons.
The design of the rtc systray dynamic component has been fixed/improved here   3f0ad5b89b84c90aa436917fb6e731e4243e6483 .

Normally we should proceed component by component, but since I changed rtc navbar icons I updated all the other rtc icons as well for consistency 78e384a40ea8054b779bda03c7e9f26bf2972875

About systray icons themselves, I did my best considering the task's scope.
The selector has been deeply simplified (a243ca889a0ebafc07ae075a63c6cb3f220cb4e3), but I can't go deeper without creating a specific component.

@sdegueldre (and @seb-odoo ), I tried community, on browserstack too,  but I still can't replicate :/
It looks fine, except a 1px gap at the very top. Is this what are you  referring to? 

![image](https://user-images.githubusercontent.com/8558351/149187088-b6009a54-c3ba-4892-aa11-8f21f80ed0aa.png)

@aab-odoo thanks :) the studio navbar is fixed too 


@qsm-odoo ,  just a reminder ;)

seb-odoo commented:
I think the mains issues I have with the counter:

1. there are 5 pixels above the number and only 3 below it
2. the bulk of the discuss icon is 3 px higher than the bottom, so it gives the feeling it is misaligned
3. the activity icon is actually misaligned because it is 1 pixel less tall than the others

![image](https://user-images.githubusercontent.com/38213504/149193150-e09cbf44-1e78-476f-89b2-54334988571b.png)

sdegueldre commented:
@stefanorigano here is what I get in enterprise (slightly off center, maybe just 1 px):
![image](https://user-images.githubusercontent.com/42469486/149279346-c3a50814-5e2a-412b-9b72-a51b6d629eb3.png)
Here is what I get in community (very off center):
![image](https://user-images.githubusercontent.com/42469486/149279384-845340a7-5aca-425b-b4f0-dc577bd882a9.png)

Here is what happens when I zoom
enterprise:
![image](https://user-images.githubusercontent.com/42469486/149279523-f74ad067-a580-4faf-905c-be93ff430ac6.png)
community:
![image](https://user-images.githubusercontent.com/42469486/149279567-d7517d4d-d0cb-419b-b823-b7f1dc418e7e.png)

As you can see the enterprise badges get *more* centered when you zoom so I think if there's anything wrong with it it's probably just a rounding problem with the browser, but that's fine, however the community version stays very off-center even when zoomed it.

The problem seems to be more pronounced on firefox than on chrome. Please note that this has nothing to do with your work as far as I can tell, there is the very same problem with reactions in the chatter in community vs enterprise. I was just wondering if you were aware of the problem and would be ok with switching to Roboto for community since that seems to fix the issue.

stefanorigano commented:
thanks @sdegueldre , 

> would be ok with switching to Roboto for community since that seems to fix the issue.

Roboto is still considered an enterprise "feature" ,  we have to deal with these kind discrepancies I guess :/
About the badges, the problem is that each os/browser renders fonts as they please, and it's even more true for tiny elements.
I may have found a solution... it costs us a couple of SCSS lines, but should be fine. 

Unfortunately neither me or my team (or browserstack) can test the linux/firefox scanario that seems the trickiest one, could you have a look? Thanks so much 🙏🏼
https://runbot.odoo.com/runbot/r-d-1?search=master-scss-revamp-navbar-sri

sdegueldre commented:
@stefanorigano doesn't seem to fix the issue, although I guess it's not a huge problem so don't worry too much about it, thanks for looking into it anyway!

stefanorigano commented:
Hello @sdegueldre , thanks for checking it out.
This is unfortunate, indeed... we'll have to investigate more I guess.
As said it's already reproducible in in current master (follows what I get with osx/chrome)... the good news is that with this PR systray's badges will appear consistently broken or fixes without difference between modules.

Thanks again!
![image](https://user-images.githubusercontent.com/8558351/149764422-4d94fbcf-6e99-4de7-86e9-44902cb74f6d.png)

qsm-odoo commented on addons/web/__manifest__.py:
```diff
@@ -99,6 +99,7 @@
             'web/static/src/search/**/*',
             'web/static/src/views/**/*',
             'web/static/src/webclient/**/*',
+            ('remove', 'web/static/src/webclient/navbar/navbar.scss'), # already in _assets_common_styles
```
I don't understand this ? It's in_ssets_common_styles indeed but that is not in the backend assets so it does not have to be removed ? :thinking: Note: I don't understand why that _assets_common_styles exists at all to be honest, it is only included once... by the common assets :shrug: 

Anyways I guess this will go away anyway since ADR is getting rid of the common assets bundle I guess.

qsm-odoo commented on addons/web/static/src/legacy/scss/ui.scss:
```diff
@@ -1,4 +1,4 @@
-.o_hidden {
+:root .o_hidden {
```
Note: if that kind of stuff is needed, it needs comments

stefanorigano replied:
sure

qsm-odoo commented on addons/website/__manifest__.py:
```diff
@@ -144,6 +144,7 @@
         ],
         'web._assets_primary_variables': [
             'website/static/src/scss/primary_variables.scss',
+            ('prepend', 'website/static/src/scss/website.navbar.variables.scss'),
```
Normally the point of having an assets splits in multiple part (variable, secondary, bootstrap, etc) is to avoid the need of prepend/append/etc: you should normally be able to just "add at the end of the parent module files". So my question is: why the need of prepend here?

stefanorigano replied:
Long story short, we decided to use `primary_variables.scss` for global variables only (eg. brand colors).
Component-related variables will be moved into each component folder, in a scss file with `.variable.scss` suffix. 

A glob fetching all these files  [has been added](https://github.com/odoo/odoo/pull/82434/files#diff-66d417d1828879bbc2124ddc118b47030ea7c30fb93e660c5231365420002214R304) in `web` manifest.
```
'web._assets_primary_variables': [
    'web/static/src/legacy/scss/primary_variables.scss',
    'web/static/src/**/**/*.variables.scss', ⬅︎
],
```
It will probably cause a bit of file juggling between files, but the approach is more consistent with the modular way we're pushing forward with owl.

cc. @aab-odoo

stefanorigano replied:
The reason was that, by default, navabar links use `$o-font-size-base` that's 13px in the backend, but 12px in the frontend.

I'm now realising that this will override the backend value too, I didn't notice that because both aim for same value, 13px.

I'll "force" website navbar links in CSS for now... it will not be pretty like a variable change, but as you said all of this will be removed soon

qsm-odoo commented on addons/website/static/src/scss/website.navbar.mobile.scss:
```diff
@@ -16,13 +16,17 @@
-        // = Invalidate top-menu native 'grid' layout
+
+        // = Invalidate top-menu native 'flex' layout
         // --------------------------------------------------------------------
-        &, .o_menu_sections li, .dropdown-menu.show {
-            display: block;
+        display: block !important;
+
+        .o_menu_toggle {
+            display: inline !important;
```
This commit modifies a file and move it at the same time: never do that, move/rename it first then modify it in another commit. Here Github is smart enough to understand it but that might not be the case of all Git configurations and that would be annoying to rebase / cherry pick.

stefanorigano replied:
Done thanks

qsm-odoo reviewed (changes requested):
Hello @stefanorigano 

Sorry it took me a while before checking this after being back from holidays. I flew over the code. My general feeling is that I will not worry too much about any change on the website navbar and frontend specific styles as this will disappear in a near future (if everything goes according to plan, so of course that's not a reason to merge a visually broken frontend but that's not the case here, so ok for me :+1:).

I added some technical comments.

stefanorigano commented on addons/website/views/website_navbar_templates.xml:
```diff
@@ -75,16 +75,16 @@
+                <div class="o_menu_systray d-none d-md-flex ml-auto" groups="website.group_website_publisher">
                     <t t-call="website.website_publisher">
-                        <t t-set="extra_classes" t-value=""/>
+                        <t t-set="extra_classes" t-value="'o_menu_systray_item'"/>
```
I didn't know, thanks 🙏🏼

stefanorigano commented on addons/website/views/website_navbar_templates.xml:
```diff
@@ -30,9 +30,9 @@
                 </p>
             </div>
-            <nav groups="base.group_user" t-if="website" id="oe_main_menu_navbar" class="o_main_navbar">
+            <nav groups="base.group_user" t-if="website" id="oe_main_menu_navbar" class="o_main_navbar d-flex align-items-center">
```
Yes, agreed. In this case it was necessary

stefanorigano commented:
@qsm-odoo , thanks so much.
I added some custom CSS to handle navbar links in website and I've split in 2 the commit that moved and edited `website.navbar.mobile.scss`.
Thanks

stefanorigano commented:
@adr-odoo , rebased on https://github.com/odoo/odoo/pull/83501
Ready and green 👍

adr-odoo reviewed (approved):
@rebase-ff r+

adr-odoo commented:
@robodoo rebase-ff r+

robodoo commented:
Merge method set to rebase and fast-forward
